packages/navigator/src/browser/navigator-search.ts

```diff
@@ -0,0 +1,222 @@
+         * no filter is set.
+         */
+        readonly filteredNodes: ReadonlyArray<Readonly<TreeNode>>;
+
+        /**
+         * Event that is fired when the filtered nodes have been changed.
+         */
+        readonly onFilteredNodesChanged: Event<ReadonlyArray<Readonly<TreeNode>>>;
+
+    }
+
+    /**
+     * Options for the search term throttle.
+     */
+    @injectable()
+    export class ThrottleOptions {
+
+        /**
+         * The delay (in milliseconds) before the throttle notifies clients about its content change.
+         */
+        readonly delay: number;
+
+    }
+
+    export namespace ThrottleOptions {
+
+        /**
+         * The default throttle option.
+         */
+        export const DEFAULT: ThrottleOptions = {
+            delay: 50
+        };
+
+    }
+
+    /**
+     * The search term throttle. It notifies clients if the underlying search term has changed after a given
+     * amount of delay.
+     */
+    @injectable()
+    export class Throttle implements Disposable {
+
+        protected readonly disposables = new DisposableCollection();
+        protected readonly emitter = new Emitter<string | undefined>();
+
+        protected timer: number | undefined;
+        protected state: string | undefined;
+
+        constructor(@inject(ThrottleOptions) protected readonly options: ThrottleOptions) {
+            this.disposables.push(this.emitter);
+        }
+
+        append(input: string | undefined): string | undefined {
+            if (input === undefined) {
+                this.reset();
+                return undefined;
+            }
+            this.clearTimer();
+            if (this.state === undefined) {
+                this.state = input;
+            } else {
+                if (input === '\b') {
+                    this.state = this.state.length === 1 ? '' : this.state.substr(0, this.state.length - 1);
+                } else {
+                    this.state += input;
+                }
+            }
+            this.timer = window.setTimeout(() => this.fireChanged(this.state), this.options.delay);
+            return this.state;
+        }
+
+        get onChanged(): Event<string | undefined> {
+            return this.emitter.event;
+        }
+
+        dispose(): void {
+            this.disposables.dispose();
+        }
+
+        protected fireChanged(value: string | undefined) {
+            this.clearTimer();
+            this.emitter.fire(value);
+        }
+
+        protected clearTimer() {
+            if (this.timer) {
+                window.clearTimeout(this.timer);
+                this.timer = undefined;
+            }
+        }
+
+        protected reset() {
+            this.state = undefined;
+            this.fireChanged(undefined);
+        }
+
+    }
+
+}
+
+export namespace FileNavigatorSearch {
+
+    @injectable()
+    export class EngineImpl implements Engine, TreeDecorator {
```

Comment:
why cannot it be top-level `FileNavigatorSearch`? without interfaces and symbols, it seems to be simpler to me

Reply:
I agree. I have changed it multiple times until I had a version that compiled and worked with inversify too. I do the refactorings right now.

Reply:
I am getting tired of getting this crap fro inversify all the time:
```
Uncaught (in promise) Error: NULL argument
    at Lookup.hasKey (lookup.js:61)
    at getBindings (planner.js:116)
    at _getActiveBindings (planner.js:30)
    at _createSubRequests (planner.js:91)
    at planner.js:108
    at Array.forEach (<anonymous>)
    at planner.js:107
    at Array.forEach (<anonymous>)
    at _createSubRequests (planner.js:94)
    at planner.js:108
```